### PR TITLE
collectIt: better collect; also closes #16078, #16098 #16372

### LINKED ISSUE #16078: `sugar.collect` doesn't work with `string` + other bugs
`sugar.collect` doesn't work with `string`
It's a special case but an important one.

### Example 1
```nim
when true:
  import std/sugar
  let s = collect(newStringOfCap(4)):
    for j in 0..<4: char(j)
```

### Current Output
`Error: cannot instantiate: 'newStringOfCap[typeof:`

### Expected Output
works

### Example 2
`collect` doesn't work with `when`
```nim
when true:
  import std/sugar
  let s = collect(newSeqOfCap(4)):
    when true:
      for j in 0..<4: char(j)
    else:
      for j in 0..<3: char(j)
  echo s
```
### Additional Information
* devel 1.5.1 d9038ed792b923cfbb9593ab82825fdb48728adc
* also reported in forum [  TITLE: one more "collect" question (bug?) - Nim forum](https://forum.nim-lang.org/t/7253)

### LINKED ISSUE #16098: collect macro should support Nim's builtin (bit-)set
```
import sugar

let BS = collect(hash[int8]) :   # or a similar syntax
  for i in 0.int8 .. 8 :
    if i mod 2 == 0 : i

echo BS
```
#[ gives
/usr/lib/nim/pure/sugar.nim(352, 14) collect
/usr/lib/nim/core/macros.nim(1091, 27) expectKind
/home/jarausch/Nim_My/BS_Collect.nim(3, 22) Error: Expected one of {nnkIdent, nnkSym, nnkCall}, got nnkBracketExpr
#]

### LINKED ISSUE #16372: collect(newString) : cannot instantiate
collect_string_bug.nim(3, 17) template/generic instantiation of `collect` from here
/usr/lib/nim/pure/sugar.nim(185, 21) Error: cannot instantiate: 
```
  for i in 0 .. 2:
    "this does not work"
  ]'; got 1 typeof(s) but expected 0
```

when trying to compile
```
import sugar

var st = collect(newString) :
  for i in 0..2 :  "this does not work"
  
echo st
```
*```
$ nim -v
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2020-12-16
Copyright (c) 2006-2020 by Andreas Rumpf

git hash: 9ef7c0182058141b81a41e32b699be8f2e42f3e9
active boot switches: -d:release

```

### PR DESCRIPTION
**Pull request by timotheecour**
TLDR: explicit `yield e` instead of implicit `e` offers more flexibility, without causing ambiguities

## `collect` vs `collectIt`:
* collectIt is more flexible, eg works with `string`
```nim
# generate a random string starting with "pre"
echo "pre".collectIt(for i in 1..4: yield sample({'a'..'z'})) # prexghh
```

* allows multiple "yield", unlike collect, eg for zipping
```nim
    let x1 = "".collectIt:
      yield '{'
      for c in "foo": yield c
      for c in 'a'..'c': yield c
      yield '}'
    doAssert x1 == "{fooabc}"
```

* doesn't have the bugs of collect mentioned in https://github.com/nim-lang/Nim/issues/16078 (eg top-level `when` breaks collect)

* allows accessing the container when needed:
```nim
   import std/sets
    let y = initHashSet[int]().collectIt:
      for a in 0..3:
        debugEcho it # this wouldn't be possible with collect, you'd have to transform to explicit block
        yield {a*2} # syntax for set
      it.excl 0 # possible too
      it.incl it.len*10 # possible too
    doAssert y == [4, 2, 6, 30].toHashSet
```

* thanks to that, it also works with array, and allows inserting out of order or other custom logic

* closes #16078 by providing a suitable alternative
* closes #16098 ditto
* closes #16372 ditto

* etc, see tests

## note
the only advantage of collect currently is type inference, in the case where it's needed. However, again in this case, explicit (via placeholders) is better than implicit and future versions will allow type inference as follows:
```nim
# newSeq[_] or newSeq[auto], not sure yet
let s2 = collectIt(newSeq[_]): for i in 0..2: yield i.float # will infer seq[float]
let s3 = collectIt(initTable[_, _]): for i in 0..2: yield {i: $i} # will infer initTable[int, string]
```
it's better because it's more flexible.

## note2:
omitting `init` will in future also be possible, infering seq[T]:
```nim
let s1 = collectIt: for i in 0..2: yield i.float # will infer seq[float] here
```

### CODE REVIEW

**Comment by planetis-m** [comment]
> the only advantage of collect currently is type inference, in the case where it's needed. 

Indeed type inference is also a disadvantage if you want to collect in a string/another container that has no generics. Never run into this problem because in case of type alias you can work with the aliased type and strings wasn't really a use case of collect to me.

But we can discuss if it's worth removing type inference and having everything explicit. Like always having to write: `collect(newSeq[T], for x in a.items: x)`. I just don't know if it's ok breaking it right now. Maybe we need to deprecate `collect` and have another name like `compose`.

About inferring the container type. We had discussions previously and came up with this design.
There is a version of [`collect`](https://github.com/planetis-m/looper) without `init` parameter and is a close design, as it can't support other "unknown" collection types. I think if `collectIt` can manage to support both explicit and implicit container types then `collect` can too. In fact I think I know a solution.

Supporting `when` was a one liner. Not a real argument.

Also Araq suggested we can make a guess about the size of the container but the implementation was too hard for me. In the end we settled with user passing a guess like `collect(initTable(50), ...`.

Using yield is a cool idea but is very verbose. It's only advantage is you can have yield that is not the last statement like your second example. Pretty obscure feature imo. ~Also `collectIt` is misleading as there is no `it`, it's more like `collectYield`.~

In general I think everything wrong with collect can be fixed, although some might require breaking changes.

**Comment by timotheecour** [comment]
> Using yield is a cool idea but is very verbose.

it fits more with rest of nim, in particular how iterators are written.

> It's only advantage is you can have yield that is not the last statement like your second example. Pretty obscure feature imo.

it enables more flexibility, eg ~~`zip`-ing~~ EDIT: chaining:

`"",collectIt(for a in x1: yield a; for b: in x2: yield b)`

or more general control flow:

`"",collectIt(for a in x1: (yield a; debugEcho "after"))`

I don't think that's obscure at all.

> Also collectIt is misleading as there is no it, it's more like collectYield.

the `it` refers to the container, which is made available, see tests. 

> although some might require breaking changes.

no breaking change needed if we add `collectIt`

**Comment by planetis-m** [comment]
That's zipping!? Real zip can only be implemented with an iterator, I know it can't work bc I saw even with yield you still needed the special syntax: x for add, {x} for incl, {a: b} for []=.

**Comment by planetis-m** [comment]
Cool effort, you have opened a discussion and that's appreciated. Things I like about it:

1. Being able to infer the container type, although that's not implemented yet.
2. Using `it` to refer to the container. Makes it possible to use arrays. Good idea.

**Comment by planetis-m** [comment]
> or more general control flow:
> 
> `"",collectIt(for a in x1: (yield a; debugEcho "after"))`
> 
> I don't think that's obscure at all.

it totally is but anyway, this Pr is superseded by #16089

**Comment by Araq** [comment]
```nim

 let y = initHashSet[int]().collectIt:
      for a in 0..3:
        debugEcho it # this wouldn't be possible with collect, you'd have to transform to explicit block
        yield {a*2} # syntax for set
      it.excl 0 # possible too
      it.incl it.len*10 # possible too
```

is just

```nim

 var y = initHashSet[int]()
 for a in 0..3:
   debugEcho y
   y.incl a * 2
   y.excl 0
   y.incl y.len*10 # possible too

```

It's already super easy to rewrite a ``collect`` expression should the need arise, ``collectIt`` doesn't seem to solve any real problem. I mean ``collect`` also does not do all that much but with ``it`` and ``yield`` it just lost any justification for its existence.
